File: Solutions/12. Integer to Roman/12-integer-to-roman.java
/*
    LEETCODE 13 - ROMAN TO INTEGER
    
    coded by Fatih Cinar
    on March 7th, 2020
*/


public class Roman
{
    private final int I = 1;
    private final int IV = 4;
    private final int V = 5;
    private final int IX = 9;
    private final int X = 10;
    private final int XL = 40;
    private final int L = 50;
    private final int XC = 90;
    private final int C = 100;
    private final int CD = 400;
    private final int D = 500;
    private final int CM = 900;
    private final int M = 1000;

    private final int[] allValues = {M, CM, D, CD, C, XC, L, XL, X, IX, V, IV, I};

    private final int numberOfAllValues = 13;



    public static void main(String[] args){
        System.out.println("Introduction");

        Roman romanConversionObject = new Roman();

        String romanForm = romanConversionObject.integerToRoman(1994);


        System.out.println(romanForm);
    }

    private String getRomanCounterpart(int singleValue){
        /*
            This method gets a single-value
            and returns its Roman counterpart
            e.g. gets 4 , returns IV string
        */

        switch (singleValue) {
            case I:
                return "I";
            case IV:
                return "IV";
            case V:
                return "V";
            case IX:
                return "IX";
            case X:
                return "X";
            case XL:
                return "XL";
            case L:
                return "L";
            case XC:
                return "XC";
            case C:
                return "C";
            case CD:
                return "CD";
            case D:
                return "D";
            case CM:
                return "CM";
            case M:
                return "M";
            default:
                return "";

        }

    }


    public String integerToRoman(int value){

        String romanForm = "";

        for(int iterator = 0; iterator < numberOfAllValues; iterator++){

            int valueNow = allValues[iterator];

            while(value >= valueNow){

                value -= valueNow;

                romanForm += getRomanCounterpart(valueNow);
            }

        }

        return romanForm;
        
    }

    
}

File: Solutions/12. Integer to Roman/12-integer-to-roman.cpp
/*
    LEETCODE 12 - INTEGER TO ROMAN

    coded by Fatih Cinar
    on March 9th, 2020
*/


#include <iostream>
#include <vector>
using namespace std;

string getRomanCounterpart(int singleValue);
    /*
        This function gets an integer number
        and returns its Roman equivalent
    */

string integerToRoman(int value);
    /*
        This function gets an integer value
        and returns the number in Roman Alphabet
    */


enum Values
{
    I = 1,
    IV = 4,
    V = 5,
    IX = 9,
    X = 10,
    XL = 40,
    L = 50,
    XC = 90,
    C = 100,
    CD = 400,
    D = 500,
    CM = 900,
    M = 1000
};

string getRomanCounterpart(int singleValue){
    /*
        This function gets an integer number
        and returns its Roman equivalent
        Only specific ones
        Kind of hash-map
    */

    switch(singleValue){
        case I:
            return "I";
        case IV:
            return "IV";
        case V:
            return "V";
        case IX:
            return "IX";
        case X:
            return "X";
        case XL:
            return "XL";
        case L:
            return "L";
        case XC:
            return "XC";
        case C:
            return "C";
        case CD:
            return "CD";
        case D:
            return "D";
        case CM:
            return "CM";
        case M:
            return "M";
        default:
            return "";

    }


}

string integerToRoman(int value){
    /*
        This function gets an integer value
        and returns the number in Roman Alphabet
    */

    const int numberOfValues = 13;
    // there are 13 values, consider like 13 digits

    int allValues[numberOfValues] =
        {M, CM, D, CD, C, XC, L, XL, X, IX, V, IV, I};
    // We will find these values in integer number
    // BE EXTREMELY CAREFUL
    // add the values to the array in descending order !!

    /*
        ALGORITHM
        # Starting with the biggest values
        # Find each value in the integer number
        # And when you find it, add it to Roman string
    */

    string romanForm = "";

    int valueNow;
    // the value dealing with now

    for (int iterator = 0; iterator < numberOfValues; iterator++)
    {

        valueNow = allValues[iterator];

        // Find the values inside the integer as much as you can
        while (value >= valueNow)
        {
            // Remove one amount from value
            // add the amount to the string
            value -= valueNow;
            romanForm += getRomanCounterpart(valueNow);
        }
    }

    return romanForm;
}

